### PR TITLE
Refactor fields

### PR DESCRIPTION
**Pull request by corranwebster**
This is a refactor of the Field class to split it into display and editable variants: base `Fields` may not be editable, `EditaleFields` add editability (ie. the value can be changed by the user).  Includes tests and documentation, including fixes to make the documentation actually compile.

Adds a LabelField and ImageField to demonstrate the non-editable Fields.

### CODE REVIEW

File: pyface/fields/i_editable_field.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    """ The editable field interface.
+
+    A editable field is a widget that displays a user-editable value.
+    """
+
+
+class MEditableField(HasTraits):
+    """The editable field mix-in.
+
+    Classes which use this mixin should implemnt _observe_control_value to
```
Comment (mdickinson) [inline pyface/fields/i_editable_field.py:29]:
```suggestion
    Classes which use this mixin should implement _observe_control_value to
```

Comment (mdickinson) [reply pyface/fields/i_editable_field.py:29]: It looks like classes using this mixin should also implement `_get_control_value` (either directly, or indirectly - e.g., through also inheriting from `MField` or `Field`). Should that also be documented here?

Would it make sense to have `MEditableField` inherit from `MField`?

File: pyface/fields/i_field.py [hunk 2]
```diff
@@ -88,13 +100,22 @@ def _set_control_value(self, value):
         """ Toolkit specific method to set the control's value. """
         raise NotImplementedError()
 
-    def _observe_control_value(self, remove=False):
-        """ Toolkit specific method to change the control value observer. """
-        raise NotImplementedError()
+    def _get_control_alignment(self):
+        """ Toolkit specific method to get the control's read_only state. """
+        raise NotImplementedError
```
Comment (mdickinson) [inline pyface/fields/i_field.py:105]: Consistency nit:
```suggestion
        raise NotImplementedError()
```

Comment (mdickinson) [inline pyface/fields/i_field.py:105]: Consistency nit:
```suggestion
        raise NotImplementedError()
```

File: pyface/fields/i_label_field.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+
+from traits.api import HasTraits, Str
+
+from pyface.fields.i_field import IField
+from pyface.ui_traits import Image
+
+
+class ILabelField(IField):
+    """ The text field interface. """
```
Comment (mdickinson) [inline pyface/fields/i_label_field.py:21]: Wording nit:
```suggestion
    """ The label field interface. """
```

File: pyface/fields/i_label_field.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+# All rights reserved.
+#
+# This software is provided without warranty under the terms of the BSD
+# license included in LICENSE.txt and may be redistributed only under
+# the conditions described in the aforementioned license. The license
+# is also available online at http://www.enthought.com/licenses/BSD.txt
+#
+# Thanks for using Enthought open source!
+
+""" The text field interface. """
```
Comment (mdickinson) [inline pyface/fields/i_label_field.py:11]:
```suggestion
""" The label field interface. """
```

File: pyface/fields/image_field.py [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+# All rights reserved.
+#
+# This software is provided without warranty under the terms of the BSD
+# license included in LICENSE.txt and may be redistributed only under
+# the conditions described in the aforementioned license. The license
+# is also available online at http://www.enthought.com/licenses/BSD.txt
+#
+# Thanks for using Enthought open source!
+
+""" The spin field widget. """
```
Comment (mdickinson) [inline pyface/fields/image_field.py:11]:
```suggestion
""" The image field widget. """
```

File: pyface/ui/qt/fields/image_field.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+from traits.api import provides
+
+from pyface.ui.qt.fields.field import Field
+from pyface.fields.i_image_field import IImageField, MImageField
+
+
+@provides(IImageField)
+class ImageField(MImageField, Field):
+    """ The Qt-specific implementation of the field class
```
Comment (mdickinson) [inline pyface/ui/qt/fields/image_field.py:23]:
```suggestion
    """ The Qt-specific implementation of the image field class
```

File: pyface/ui/qt/fields/image_field.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+    This is an abstract class which is not meant to be instantiated.  Because
+    many concrete QWidgets provide a `value` property, the default getters and
+    setters target this.
```
Comment (mdickinson) [inline pyface/ui/qt/fields/image_field.py:27]: I think this text can be removed.

File: pyface/ui/qt/fields/toggle_field.py [hunk 8]
```diff
@@ -60,6 +60,16 @@ def _observe_control_value(self, remove=False):
...
 
+    def _get_control_alignment(self):
+        """ Toolkit specific method to get the control's alignment. """
+        # dummy implementation
+        return self.alignment
+
+    def _set_control_alignment(self, alignment):
+        """ Toolkit specific method to set the control's alignment. """
+        # use stylesheet for button alignment
+        self.control.setStyleSheet(f"text-align: {alignment}")
```
Comment (mdickinson) [inline pyface/ui/qt/fields/toggle_field.py:71]: Would this remove any previous style-sheet modifications that are unrelated to alignment?

File: pyface/ui/qt/util/alignment.py [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+}
+ALIGNMENT_MASK = (
+    Qt.AlignmentFlag.AlignLeft
+    | Qt.AlignmentFlag.AlignHCenter
+    | Qt.AlignmentFlag.AlignRight
+)
+
+
+def alignment_to_qalignment(alignment):
+    return ALIGNMENT_TO_QALIGNMENT[alignment]
```
Comment (mdickinson) [inline pyface/ui/qt/util/alignment.py:36]: It would be good to have docstrings (or maybe just type annotations) here.

File: pyface/ui/wx/fields/editable_field.py [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+
+from traits.api import provides
+
+from pyface.fields.i_editable_field import IEditableField, MEditableField
+from pyface.ui.wx.fields.field import Field
+
+
+@provides(IEditableField)
+class EditableField(MEditableField, Field):
+    """ The Wx-specific implementation of the Editablefield class
```
Comment (mdickinson) [inline pyface/ui/wx/fields/editable_field.py:22]:
```suggestion
    """ The Wx-specific implementation of the EditableField class
```

File: pyface/ui/wx/fields/image_field.py [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+from traits.api import provides
+
+from pyface.ui.wx.fields.field import Field
+from pyface.fields.i_image_field import IImageField, MImageField
+from pyface.wx.image_control import ImageControl
+
+
+@provides(IImageField)
+class ImageField(MImageField, Field):
+    """ The Wx-specific implementation of the field class
```
Comment (mdickinson) [inline pyface/ui/wx/fields/image_field.py:22]:
```suggestion
    """ The Wx-specific implementation of the image field class
```

File: pyface/ui/wx/fields/image_field.py [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+
+    This is an abstract class which is not meant to be instantiated.  Because
+    many concrete QWidgets provide a `value` property, the default getters and
+    setters target this.
```
Comment (mdickinson) [inline pyface/ui/wx/fields/image_field.py:26]: Suggest removing:
```suggestion
```

File: pyface/wx/image_control.py [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+# Thanks for using Enthought open source!
+
+""" Defines a wxPython ImageControl widget that is used by various Widgets.
+"""
+
+
+import wx
+
+
+class ImageControl(wx.Window):
```
Comment (mdickinson) [inline pyface/wx/image_control.py:18]: If I wanted to see this code in action, what's the easiest way for me to exercise this code manually? Are there examples that make use of this?

**Review by mdickinson — APPROVED** [review]
Done with a (not very thorough) review. Amongst other comments, there are some copy-paste errors, and a typo in `pyface.fields.api` that breaks `from pyface.fields.api import EditableField`.

File: pyface/fields/i_editable_field.py (hunk 1 above)
Comment (corranwebster) [reply pyface/fields/i_editable_field.py:29]: `Field` requires `_get_control_value` (it's used for testing - we need a way in tests to see what value is set), so it should

Regarding the inheritance, I've been trying to keep the mixins from inherting from each other.  IIRC a few years ago I ran into an issue where Trait declarations were wrong when using inheritance in these mixins because trait definitions aren't evaluated following `mro`. (This is the relevant issue, I think https://github.com/enthought/traits/issues/401)

File: pyface/fields/api.py [hunk 14]
```diff
@@ -14,6 +14,8 @@
 
 - :class:`~.CheckBoxField`
 - :class:`~.ComboField`
+- :class:`~.ImageField`
+- :class:`~.LabelField`
```
[the remark this answers is not in the document]
Comment (corranwebster) [reply pyface/fields/api.py:18]: Yes!

File: pyface/ui/qt/fields/toggle_field.py (hunk 8 above)
Comment (corranwebster) [reply pyface/ui/qt/fields/toggle_field.py:71]: Yes, for widget-level style-sheet modifications but not application-wide style-sheets, I think.  I'm OK with this because Pyface doesn't expose the style-sheets anywhere else and anyone who is dropping down to the Qt level for a Field should be aware that they may be stepping on the toes of the field infrastructure.

If we find that we need to set other characteristics via style sheets we may need to add a layer to manage merging stylesheet settings, but that would likely be at the level of the Widget.  So for a future PR if it is an issue

File: pyface/wx/image_control.py (hunk 13 above)
Comment (corranwebster) [reply pyface/wx/image_control.py:18]: No examples currently, but this was copied and simplified from the wx TriatsUI ImageEditor's ImageControl: https://github.com/enthought/traitsui/blob/main/traitsui/wx/image_control.py

Comment (corranwebster) [reply pyface/wx/image_control.py:18]: You could exercise it with something like:
```
import wx
from pyface.api import ImageResource
from pyface.wx.image_control import ImageControl

window = wx.Frame()
# hack to get a wx bitmap - could also get it through wx api and an image file
bitmap = ImageResource('splash').create_bitmap()
control = ImageControl(window, bitmap)
window.Show()
```
I think (my wx is rusty)

File: pyface/fields/i_field.py [hunk 15]
```diff
@@ -88,13 +100,22 @@ def _set_control_value(self, value):
...
+
+    def _set_control_alignment(self, alignment):
+        """ Toolkit specific method to set the control's alignment. """
+        raise NotImplementedError
```
Comment (corranwebster) [inline pyface/fields/i_field.py:109]:
```suggestion
        raise NotImplementedError()
```

File: pyface/ui/qt/fields/image_field.py (hunk 7 above)
Comment (corranwebster) [inline pyface/ui/qt/fields/image_field.py:27]:
```suggestion
```

**Comment by corranwebster** [comment]
This is probably going to fail because of PySide 6.5
